### DISCUSSION TITLE
Anticollision Algorithm

### DISCUSSION THREAD

**Discussion by danielgarthur** [neanes/neanes#262] [Ideas]
The font currently uses OpenType features to attempt to prevent common collisions between marks (gorgons, bar lines, ison indicators, etc.), but the number of possibilities is too great to handle in a manual fashion. Currently, the burden is on the user to reposition colliding neumes.

The font metadata file already contains all of the bounding boxes for each glyph (in `glyphBBoxes`), as well as all the anchors for each mark (in `glyphsWithAnchors`). With this information, it should be possible to create or adapt an existing anticollision algorithm to automatically re-position the glyphs with Neanes as glyphs are added. 

Simply speaking, this algorithm would need to move apart rectangles to prevent overlaps. It would perhaps need to be a little smarter, in that it may need to have suggestions on which directions to move certain types of glyphs, or to give priority to certain glyphs (i.e. move these glyphs before attempting to move these other glyphs).

Users should have the option to turn this feature on or off, and the ability to reposition manually should remain.

**Comment by basil** [comment]
I had a similar thought two weeks ago. [This page](https://stackoverflow.com/questions/3265986/an-algorithm-to-space-out-overlapping-rectangles) discusses some general solutions to the problem. In general the layout engine needs a lot of work to produce more professional-looking output, such as a better line breaking algorithm.

To what degree would such an approach obsolete the mark-to-base positioning design of SBMuFL? That is, if starting from scratch would we want to have the consumer do all positioning, or is there still some intrinsic value in defining some of the positioning in the font?

**Comment by danielgarthur** [comment]
The mark-to-base positioning would still be useful to give the initial, default positions of marks.

Something else I was thinking is that I'm not sure how to avoid collisions with the base neumes themselves, unless more bounding boxes are defined. E.g.
<img width="68" alt="image" src="https://github.com/neanes/neanes/assets/4132779/7d34d90b-eae8-4142-8a86-0a68401d1179">
Or even more precisely
<img width="70" alt="image" src="https://github.com/neanes/neanes/assets/4132779/07918d22-81ab-45d5-aa4a-daba9a0b9caf">

I'm not sure how to generate such boxes programmatically, although certainly they could be created by hand and added to the metadata.

**Comment by basil** [comment]
Perhaps a text shaping engine like [HarfBuzz](https://github.com/harfbuzz/harfbuzz) could be used to achieve this goal. It allows you to [write your own shaping engine in WebAssembly and embed it into the font](https://github.com/harfbuzz/harfbuzz/blob/main/docs/wasm-shaper.md), so one could possibly implement an anticollision this way such that it applies to both Neanes and ByzHtml.
